crates/sqlexec/src/context.rs

```diff
@@ -490,16 +490,44 @@ impl SessionContext {
     pub async fn drop_tables(&mut self, plan: DropTables) -> Result<()> {
         let mut drops = Vec::with_capacity(plan.names.len());
         for r in plan.names {
-            let (_, schema, name) = self.resolve_table_ref(r)?;
+            let (database, schema, name) = self.resolve_table_ref(r)?;
+
+            let table_entry = match self
+                .metastore_catalog
+                .resolve_entry(&database, &schema, &name)
+            {
+                Some(CatalogEntry::Table(table)) => match &table.options {
+                    TableOptions::Internal(_) => table,
+                    other => {
+                        return Err(ExecError::Internal(format!(
+                            "Unexpected set of table options: {:?}",
+                            other
+                        )))
+                    }
+                },
+                Some(other) => {
+                    return Err(ExecError::Internal(format!(
+                        "Unexpected catalog entry type: {:?}",
+                        other
+                    )))
+                }
+                None => {
+                    return Err(ExecError::Internal(
+                        "Missing table in the catalog".to_string(),
+                    ));
+                }
+            };
+
+            let tracker = BackgroundJobDeleteTable::new(self.tables.clone(), table_entry.clone());
+            self.background_jobs.add(tracker)?;
```

Comment:
Let's replace this with an `if let ...`. This will incorrectly error when trying to drop an external table, or when running `DROP TABLE IF EXISTS <table>`.

So something like:

```rust
if let Some(CatalogEntry::Table(ent)) = self
    .metastore_catalog
    .resolve_entry(&database, &schema, &name)
{
    // We should only try to delete data for internal tables.
    if matches!(ent.options, TableOptions::Internal(_)) {
        let tracker = BackgroundJobDeleteTable::new(self.tables.clone(), ent.clone());
        self.background_jobs.add(tracker)?;
    }
}
```